File: src/daz_agent_sdk/config_test.py
from __future__ import annotations

from pathlib import Path

import yaml

from daz_agent_sdk.types import Tier
from daz_agent_sdk.config import (
    Config,
    get_image_steps,
    get_provider_config,
    get_tier_chain,
    load_config,
)


# ##################################################################
# default config — no file
# verify that load_config returns sensible defaults when no file exists
def test_default_config_no_file(tmp_path: Path) -> None:
    missing = tmp_path / "does_not_exist.yaml"
    cfg = load_config(missing)
    assert isinstance(cfg, Config)


# ##################################################################
# default tier chains
# verify all five tiers have non-empty chains with correct first entry
def test_default_tier_very_high(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    chain = cfg.tiers["very_high"].chain
    assert len(chain) >= 1
    assert chain[0] == "claude:claude-opus-4-6"


def test_default_tier_high(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    chain = cfg.tiers["high"].chain
    assert len(chain) >= 1
    assert chain[0] == "claude:claude-opus-4-6"


def test_default_tier_medium(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    chain = cfg.tiers["medium"].chain
    assert chain[0] == "claude:claude-sonnet-4-6"


def test_default_tier_low(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    chain = cfg.tiers["low"].chain
    assert chain[0] == "claude:claude-haiku-4-5-20251001"


def test_default_tier_free_fast(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    chain = cfg.tiers["free_fast"].chain
    assert chain[0] == "ollama:qwen3-8b"


def test_default_tier_free_thinking(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    chain = cfg.tiers["free_thinking"].chain
    assert chain[0] == "ollama:qwen3-30b-32k"


# ##################################################################
# default image config
# verify default model and step counts for each image tier
def test_default_image_model(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    assert cfg.image.model == "z-image-turbo"


def test_default_image_steps_very_high(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    assert cfg.image.tiers["very_high"].steps == 32


def test_default_image_steps_high(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    assert cfg.image.tiers["high"].steps == 16


def test_default_image_steps_medium(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    assert cfg.image.tiers["medium"].steps == 8


def test_default_image_steps_low(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    assert cfg.image.tiers["low"].steps == 2


# ##################################################################
# default provider config
# verify default provider entries are present
def test_default_provider_claude(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    assert "claude" in cfg.providers
    assert cfg.providers["claude"]["permission_mode"] == "bypassPermissions"


def test_default_provider_ollama(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    assert cfg.providers["ollama"]["base_url"] == "http://localhost:11434"


def test_default_provider_no_vllm(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    assert "vllm" not in cfg.providers


# ##################################################################
# default logging config
# verify default log directory, level and retention
def test_default_logging(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    assert cfg.logging.level == "info"
    assert cfg.logging.retention_days == 30
    assert "logs" in cfg.logging.directory


# ##################################################################
# default fallback config
# verify default fallback strategies
def test_default_fallback_single_shot(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    assert cfg.fallback.single_shot.strategy == "immediate_cascade"


def test_default_fallback_conversation(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    assert cfg.fallback.conversation.strategy == "backoff_then_cascade"
    assert cfg.fallback.conversation.max_backoff_seconds == 60
    assert cfg.fallback.conversation.summarise_with == "free_thinking"


# ##################################################################
# custom config overrides
# write a real yaml file and verify overrides take effect
def test_custom_tier_chain(tmp_path: Path) -> None:
    config_file = tmp_path / "config.yaml"
    data = {
        "tiers": {
            "high": ["gemini:gemini-2.5-pro", "claude:claude-opus-4-6"],
        }
    }
    config_file.write_text(yaml.dump(data), encoding="utf-8")
    cfg = load_config(config_file)
    assert cfg.tiers["high"].chain == ["gemini:gemini-2.5-pro", "claude:claude-opus-4-6"]
    # other tiers fall back to defaults
    assert cfg.tiers["low"].chain[0] == "claude:claude-haiku-4-5-20251001"


def test_custom_image_model(tmp_path: Path) -> None:
    config_file = tmp_path / "config.yaml"
    data = {
        "image": {
            "model": "dall-e-3",
            "tiers": {
                "high": {"steps": 20},
                "medium": {"steps": 10},
                "low": {"steps": 2},
            },
        }
    }
    config_file.write_text(yaml.dump(data), encoding="utf-8")
    cfg = load_config(config_file)
    assert cfg.image.model == "dall-e-3"
    assert cfg.image.tiers["high"].steps == 20
    assert cfg.image.tiers["medium"].steps == 10
    assert cfg.image.tiers["low"].steps == 2


def test_custom_provider_config(tmp_path: Path) -> None:
    config_file = tmp_path / "config.yaml"
    data = {
        "providers": {
            "ollama": {"base_url": "http://10.0.0.5:11434"},
        }
    }
    config_file.write_text(yaml.dump(data), encoding="utf-8")
    cfg = load_config(config_file)
    assert cfg.providers["ollama"]["base_url"] == "http://10.0.0.5:11434"
    # non-overridden providers still get defaults
    assert cfg.providers["claude"]["permission_mode"] == "bypassPermissions"


def test_custom_logging_config(tmp_path: Path) -> None:
    config_file = tmp_path / "config.yaml"
    data = {
        "logging": {
            "level": "debug",
            "retention_days": 7,
            "directory": "/var/log/agent-sdk",
        }
    }
    config_file.write_text(yaml.dump(data), encoding="utf-8")
    cfg = load_config(config_file)
    assert cfg.logging.level == "debug"
    assert cfg.logging.retention_days == 7
    assert cfg.logging.directory == "/var/log/agent-sdk"


def test_custom_fallback_config(tmp_path: Path) -> None:
    config_file = tmp_path / "config.yaml"
    data = {
        "fallback": {
            "single_shot": {"strategy": "retry_then_cascade"},
            "conversation": {
                "strategy": "backoff_then_cascade",
                "max_backoff_seconds": 120,
                "summarise_with": "medium",
            },
        }
    }
    config_file.write_text(yaml.dump(data), encoding="utf-8")
    cfg = load_config(config_file)
    assert cfg.fallback.single_shot.strategy == "retry_then_cascade"
    assert cfg.fallback.conversation.max_backoff_seconds == 120
    assert cfg.fallback.conversation.summarise_with == "medium"


# ##################################################################
# get tier chain function
# verify correct chain returned for each tier enum value
def test_get_tier_chain_high(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    chain = get_tier_chain(Tier.HIGH, cfg)
    assert chain[0] == "claude:claude-opus-4-6"
    assert len(chain) == 3


def test_get_tier_chain_medium(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    chain = get_tier_chain(Tier.MEDIUM, cfg)
    assert chain[0] == "claude:claude-sonnet-4-6"


def test_get_tier_chain_low(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    chain = get_tier_chain(Tier.LOW, cfg)
    assert chain[0] == "claude:claude-haiku-4-5-20251001"
    assert len(chain) == 3


def test_get_tier_chain_free_fast(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    chain = get_tier_chain(Tier.FREE_FAST, cfg)
    assert chain[0] == "ollama:qwen3-8b"


def test_get_tier_chain_free_thinking(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    chain = get_tier_chain(Tier.FREE_THINKING, cfg)
    assert chain[0] == "ollama:qwen3-30b-32k"


def test_get_tier_chain_returns_copy(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    chain1 = get_tier_chain(Tier.HIGH, cfg)
    chain1.clear()
    chain2 = get_tier_chain(Tier.HIGH, cfg)
    assert len(chain2) > 0


# ##################################################################
# get image steps function
# verify correct step count for each tier
def test_get_image_steps_very_high(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    assert get_image_steps(Tier.VERY_HIGH, cfg) == 32


def test_get_image_steps_high(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    assert get_image_steps(Tier.HIGH, cfg) == 16


def test_get_image_steps_medium(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    assert get_image_steps(Tier.MEDIUM, cfg) == 8


def test_get_image_steps_low(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    assert get_image_steps(Tier.LOW, cfg) == 2


def test_get_image_steps_unknown_tier_returns_default(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    # FREE_FAST has no image tier config — should return a sensible default
    steps = get_image_steps(Tier.FREE_FAST, cfg)
    assert isinstance(steps, int)
    assert steps > 0


# ##################################################################
# get provider config function
# verify correct dict returned for known and unknown providers
def test_get_provider_config_claude(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    pc = get_provider_config("claude", cfg)
    assert pc["permission_mode"] == "bypassPermissions"


def test_get_provider_config_ollama(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    pc = get_provider_config("ollama", cfg)
    assert pc["base_url"] == "http://localhost:11434"


def test_get_provider_config_gemini(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    pc = get_provider_config("gemini", cfg)
    assert pc["api_key_env"] == "GEMINI_API_KEY"


def test_get_provider_config_unknown_returns_empty(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    pc = get_provider_config("nonexistent_provider", cfg)
    assert pc == {}


def test_get_provider_config_returns_copy(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    pc1 = get_provider_config("ollama", cfg)
    pc1["base_url"] = "mutated"
    pc2 = get_provider_config("ollama", cfg)
    assert pc2["base_url"] == "http://localhost:11434"


# ##################################################################
# invalid yaml handling
# verify that malformed yaml falls back to defaults rather than crashing
def test_invalid_yaml_uses_defaults(tmp_path: Path) -> None:
    config_file = tmp_path / "config.yaml"
    config_file.write_text("this: is: not: valid: yaml: :::::", encoding="utf-8")
    cfg = load_config(config_file)
    assert isinstance(cfg, Config)
    assert "high" in cfg.tiers
    assert cfg.image.model == "z-image-turbo"


def test_yaml_with_null_root_uses_defaults(tmp_path: Path) -> None:
    config_file = tmp_path / "config.yaml"
    config_file.write_text("null\n", encoding="utf-8")
    cfg = load_config(config_file)
    assert isinstance(cfg, Config)
    assert cfg.tiers["high"].chain[0] == "claude:claude-opus-4-6"


def test_yaml_with_empty_file_uses_defaults(tmp_path: Path) -> None:
    config_file = tmp_path / "config.yaml"
    config_file.write_text("", encoding="utf-8")
    cfg = load_config(config_file)
    assert isinstance(cfg, Config)
    assert "medium" in cfg.tiers


# ##################################################################
# caching behaviour
# verify load_config caches when no explicit path is given
def test_force_reload_returns_fresh_config(tmp_path: Path) -> None:
    config_file = tmp_path / "config.yaml"
    data1 = {"tiers": {"high": ["gemini:gemini-2.5-pro"]}}
    config_file.write_text(yaml.dump(data1), encoding="utf-8")
    cfg1 = load_config(config_file)
    assert cfg1.tiers["high"].chain == ["gemini:gemini-2.5-pro"]

    # modify file and reload with same explicit path — explicit path is never cached
    data2 = {"tiers": {"high": ["ollama:qwen3-8b"]}}
    config_file.write_text(yaml.dump(data2), encoding="utf-8")
    cfg2 = load_config(config_file)
    assert cfg2.tiers["high"].chain == ["ollama:qwen3-8b"]


# ##################################################################
# tts config
# verify default voices are populated
def test_default_tts_voices(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    assert "gary" in cfg.tts.voices
    assert "aiden" in cfg.tts.voices
    assert cfg.tts.voices["gary"].provider == "local"


def test_default_tts_chain(tmp_path: Path) -> None:
    cfg = load_config(tmp_path / "absent.yaml")
    assert cfg.tts.default == ["local:qwen3-tts"]


def test_custom_tts_voices(tmp_path: Path) -> None:
    config_file = tmp_path / "config.yaml"
    data = {
        "tts": {
            "default": ["local:qwen3-tts"],
            "voices": {
                "alice": {"provider": "local", "voice_id": "alice", "description": "Friendly"},
            },
        }
    }
    config_file.write_text(yaml.dump(data), encoding="utf-8")
    cfg = load_config(config_file)
    assert "alice" in cfg.tts.voices
    assert cfg.tts.voices["alice"].description == "Friendly"

File: go/config_test.go
package dazagentsdk

import (
	"os"
	"path/filepath"
	"testing"
)

func TestLoadConfigDefaults(t *testing.T) {
	// Load from a nonexistent path to get pure defaults
	cfg, err := LoadConfig("/nonexistent/path/config.yaml")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	// Verify all 6 tiers exist
	expectedTiers := []string{"very_high", "high", "medium", "low", "free_fast", "free_thinking"}
	for _, tier := range expectedTiers {
		if _, ok := cfg.Tiers[tier]; !ok {
			t.Errorf("missing default tier %q", tier)
		}
	}
}

func TestDefaultTierChains(t *testing.T) {
	cfg, err := LoadConfig("/nonexistent/path/config.yaml")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	tests := []struct {
		tier     Tier
		wantLen  int
		wantFirst string
	}{
		{TierVeryHigh, 3, "claude:claude-opus-4-6"},
		{TierHigh, 3, "claude:claude-opus-4-6"},
		{TierMedium, 3, "claude:claude-sonnet-4-6"},
		{TierLow, 3, "claude:claude-haiku-4-5-20251001"},
		{TierFreeFast, 1, "ollama:qwen3-8b"},
		{TierFreeThinking, 2, "ollama:qwen3-30b-32k"},
	}

	for _, tt := range tests {
		chain := GetTierChain(tt.tier, cfg)
		if len(chain) != tt.wantLen {
			t.Errorf("GetTierChain(%v) len = %d, want %d", tt.tier, len(chain), tt.wantLen)
			continue
		}
		if chain[0] != tt.wantFirst {
			t.Errorf("GetTierChain(%v)[0] = %q, want %q", tt.tier, chain[0], tt.wantFirst)
		}
	}
}

func TestDefaultImageSteps(t *testing.T) {
	cfg, err := LoadConfig("/nonexistent/path/config.yaml")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	tests := []struct {
		tier      Tier
		wantSteps int
	}{
		{TierVeryHigh, 8},
		{TierHigh, 3},
		{TierMedium, 3},
		{TierLow, 2},
		// free_fast has no image tier, falls back to medium
		{TierFreeFast, 3},
	}

	for _, tt := range tests {
		steps := GetImageSteps(tt.tier, cfg)
		if steps != tt.wantSteps {
			t.Errorf("GetImageSteps(%v) = %d, want %d", tt.tier, steps, tt.wantSteps)
		}
	}
}

func TestDefaultImageModel(t *testing.T) {
	cfg, err := LoadConfig("/nonexistent/path/config.yaml")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if cfg.Image.Model != "z-image-turbo" {
		t.Errorf("Image.Model = %q, want %q", cfg.Image.Model, "z-image-turbo")
	}
}

func TestDefaultProviders(t *testing.T) {
	cfg, err := LoadConfig("/nonexistent/path/config.yaml")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	expectedProviders := []string{"claude", "codex", "gemini", "ollama"}
	for _, p := range expectedProviders {
		if _, ok := cfg.Providers[p]; !ok {
			t.Errorf("missing default provider %q", p)
		}
	}

	// Check claude has bypassPermissions
	if cfg.Providers["claude"]["permission_mode"] != "bypassPermissions" {
		t.Errorf("claude permission_mode = %v, want bypassPermissions", cfg.Providers["claude"]["permission_mode"])
	}

	// Check ollama has base_url
	if cfg.Providers["ollama"]["base_url"] != "http://localhost:11434" {
		t.Errorf("ollama base_url = %v, want http://localhost:11434", cfg.Providers["ollama"]["base_url"])
	}
}

func TestDefaultLogging(t *testing.T) {
	cfg, err := LoadConfig("/nonexistent/path/config.yaml")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if cfg.Logging.Directory != "~/.daz-agent-sdk/logs" {
		t.Errorf("Logging.Directory = %q", cfg.Logging.Directory)
	}
	if cfg.Logging.Level != "info" {
		t.Errorf("Logging.Level = %q", cfg.Logging.Level)
	}
	if cfg.Logging.RetentionDays != 30 {
		t.Errorf("Logging.RetentionDays = %d", cfg.Logging.RetentionDays)
	}
}

func TestDefaultFallback(t *testing.T) {
	cfg, err := LoadConfig("/nonexistent/path/config.yaml")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if cfg.Fallback.SingleShot.Strategy != "immediate_cascade" {
		t.Errorf("Fallback.SingleShot.Strategy = %q", cfg.Fallback.SingleShot.Strategy)
	}
	if cfg.Fallback.Conversation.Strategy != "backoff_then_cascade" {
		t.Errorf("Fallback.Conversation.Strategy = %q", cfg.Fallback.Conversation.Strategy)
	}
	if cfg.Fallback.Conversation.MaxBackoffSeconds != 60 {
		t.Errorf("Fallback.Conversation.MaxBackoffSeconds = %d", cfg.Fallback.Conversation.MaxBackoffSeconds)
	}
	if cfg.Fallback.Conversation.SummariseWith != "free_thinking" {
		t.Errorf("Fallback.Conversation.SummariseWith = %q", cfg.Fallback.Conversation.SummariseWith)
	}
}

func TestDefaultTTS(t *testing.T) {
	cfg, err := LoadConfig("/nonexistent/path/config.yaml")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if len(cfg.TTS.Default) != 1 || cfg.TTS.Default[0] != "local:qwen3-tts" {
		t.Errorf("TTS.Default = %v, want [local:qwen3-tts]", cfg.TTS.Default)
	}
	if len(cfg.TTS.Voices) != 2 {
		t.Errorf("TTS.Voices len = %d, want 2", len(cfg.TTS.Voices))
	}
	gary, ok := cfg.TTS.Voices["gary"]
	if !ok {
		t.Fatal("missing gary voice")
	}
	if gary.Description != "British newsreader" {
		t.Errorf("gary description = %q", gary.Description)
	}
}

func TestDefaultTransparentPostProcess(t *testing.T) {
	cfg, err := LoadConfig("/nonexistent/path/config.yaml")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if cfg.Image.TransparentPostProcess != "birefnet" {
		t.Errorf("Image.TransparentPostProcess = %q, want birefnet", cfg.Image.TransparentPostProcess)
	}
}

func TestCustomYAMLLoading(t *testing.T) {
	dir := t.TempDir()
	configFile := filepath.Join(dir, "config.yaml")

	yamlContent := `
tiers:
  high:
    - custom:model-a
    - custom:model-b
  low:
    - ollama:tiny

providers:
  custom:
    api_key: "test-key"

image:
  model: "custom-model"
  tiers:
    high:
      steps: 10
  transparent:
    post_process: "none"

logging:
  level: "debug"
  retention_days: 7
`
	if err := os.WriteFile(configFile, []byte(yamlContent), 0644); err != nil {
		t.Fatalf("failed to write config: %v", err)
	}

	cfg, err := LoadConfig(configFile)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	// Custom tier overrides default
	highChain := GetTierChain(TierHigh, cfg)
	if len(highChain) != 2 || highChain[0] != "custom:model-a" {
		t.Errorf("high chain = %v, want [custom:model-a custom:model-b]", highChain)
	}

	// Default tiers still present for ones not overridden
	mediumChain := GetTierChain(TierMedium, cfg)
	if len(mediumChain) != 3 {
		t.Errorf("medium chain should have defaults, got len %d", len(mediumChain))
	}

	// Custom provider present
	if cfg.Providers["custom"]["api_key"] != "test-key" {
		t.Errorf("custom provider api_key = %v", cfg.Providers["custom"]["api_key"])
	}

	// Default providers still present
	if _, ok := cfg.Providers["claude"]; !ok {
		t.Error("default claude provider should still be present")
	}

	// Custom image model
	if cfg.Image.Model != "custom-model" {
		t.Errorf("Image.Model = %q, want custom-model", cfg.Image.Model)
	}

	// Custom image steps for high
	if GetImageSteps(TierHigh, cfg) != 10 {
		t.Errorf("custom high steps = %d, want 10", GetImageSteps(TierHigh, cfg))
	}

	// Default image steps still present for medium
	if GetImageSteps(TierMedium, cfg) != 3 {
		t.Errorf("default medium steps = %d, want 3", GetImageSteps(TierMedium, cfg))
	}

	// Custom transparent post process
	if cfg.Image.TransparentPostProcess != "none" {
		t.Errorf("TransparentPostProcess = %q, want none", cfg.Image.TransparentPostProcess)
	}

	// Custom logging
	if cfg.Logging.Level != "debug" {
		t.Errorf("Logging.Level = %q, want debug", cfg.Logging.Level)
	}
	if cfg.Logging.RetentionDays != 7 {
		t.Errorf("Logging.RetentionDays = %d, want 7", cfg.Logging.RetentionDays)
	}
}

func TestGetTierChainUnknownTier(t *testing.T) {
	cfg, err := LoadConfig("/nonexistent/path/config.yaml")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	chain := GetTierChain(Tier("nonexistent"), cfg)
	if chain != nil {
		t.Errorf("expected nil for unknown tier, got %v", chain)
	}
}

func TestGetImageStepsUnknownTier(t *testing.T) {
	cfg, err := LoadConfig("/nonexistent/path/config.yaml")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	// free_thinking has no image tier, should fall back to medium (3)
	steps := GetImageSteps(TierFreeThinking, cfg)
	if steps != 3 {
		t.Errorf("GetImageSteps(free_thinking) = %d, want 3 (medium fallback)", steps)
	}
}

func TestGetTierChainReturnsCopy(t *testing.T) {
	cfg, err := LoadConfig("/nonexistent/path/config.yaml")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	chain1 := GetTierChain(TierHigh, cfg)
	chain2 := GetTierChain(TierHigh, cfg)

	// Mutate chain1, chain2 should be unaffected
	chain1[0] = "modified"
	if chain2[0] == "modified" {
		t.Error("GetTierChain should return a copy, not a reference")
	}
}
